Add "View Autofixes" feature for variant analysis results

jcogs33 opened the pull request:
<!-- Thank you for submitting a pull request. Please read our pull request guidelines before
  submitting your pull request:
  https://github.com/github/vscode-codeql/blob/main/CONTRIBUTING.md#submitting-a-pull-request.
-->

## Description

Adds a "View Autofixes" feature for variant analysis results. This is a canary-only feature.
Generates autofixes for selected variant analysis results and opens the resulting autofix outputs in a markdown file.
Activated via the "View Autofixes" button in the variant analysis webview or via right-clicking the variant analysis query history item. 

<img width="1068" alt="AutofixInMrvaScreenshot" src="https://github.com/user-attachments/assets/d6cd9f7d-32c6-4fc4-9ae6-668c6a8db5b3" />

</br>
</br>
At a high-level, the implementation:

- Checks for a local checkout of autofix
- Sets up to run autofix by:
  - Overriding the relevant query help
  - Getting the relevant SARIF(s)
  - Downloading source root(s)
- Runs autofix locally
- Outputs autofix results in a markdown file

Commit-by-commit review recommended. Let me know if you want me to split anything onto a separate PR.

## Consideration

- See in-line comments
- Change from my original implementation: Instead of downloading databases and then extracting the source root from the database, I directly download the source root since it seemed a bit faster. Let me know if you think the database approach is better.
- I will make a follow-up PR to convert this to support Go autofix instead of cocofix, and I'll update the `queryHelpOverrideDirectory` to the new `go/pkg/autofix/prompt/qhelps` directory in that PR.

## Testing
- I've manually tested this feature but have not added unit/view/integration tests since it seems [another canary feature](https://github.com/github/vscode-codeql/pull/3843) did not add tests. Let me know if I should add tests after all.
- Note: There is a known issue with the E2E CI test failing. It is not caused by changes in this PR.

jcogs33 commented on extensions/ql-vscode/src/common/interface-types.ts:
```diff
@@ -151,10 +151,13 @@ interface SetStateMsg {
   /** Whether to display links to the dataflow models that generated particular nodes in a flow path. */
   shouldShowProvenance: boolean;
+  /** Whether to display the "View Autofixes" button. */
+  shouldShowViewAutofixesBtn: boolean;
```
Let me know if I should make `shouldShowViewAutofixesBtn` and `shouldShowProvenance` optional. It feels a bit ugly that I am setting these values for views where they are not relevant.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/variant-analysis-view.ts:
```diff
@@ -46,13 +48,48 @@ export class VariantAnalysisView
 
     this.dataFlowPathsView = new DataFlowPathsView(app);
+
+    // Set up configuration change listener
+    this.push(
+      workspace.onDidChangeConfiguration(
+        this.onConfigurationChanged.bind(this),
+      ),
+    );
+  }
+
+  /**
+   * Handler for configuration changes
+   */
+  private onConfigurationChanged(e: ConfigurationChangeEvent): void {
+    // Check if the canary setting has changed
+    if (e.affectsConfiguration("codeQL.canary")) {
+      void this.updateUserSettings();
+    }
```
This configuration listener code was suggested by Copilot. I'm wondering if I should be using a pre-existing config listener like `VariantAnalysisConfigListener` instead? Or not bother with a config listener at all and just require closing and re-opening the variant analysis panel for the canary setting update to take affect (which is how the `shouldShowProvenance` setting works)?

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+  filterSort: RepositoriesFilterSortStateWithIds = defaultFilterSortState,
+  variantAnalyses: Map<number, VariantAnalysis>,
+  credentials: Credentials,
+  logger: NotificationLogger,
```
In retrospect, I'm not sure if I should be using `NotificationLogger`. 🤔 Let me know if I should use `extLogger` instead of or in addition to `NotificationLogger`?
Also let me know if I should add more logging in general?

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+ * @throws Error if the AUTOFIX_PATH environment variable is not set or the path does not exist.
+ */
+function findLocalAutofix(): string {
+  const localAutofixPath = process.env.AUTOFIX_PATH;
```
I'm currently requiring that users set this environment variable (see corresponding documentation). Let me know if there is something better I should do (config file, etc.) instead of using this environment variable.

koesie10 replied:
I think using a config setting would make sense (something like `codeQL.autofix.path`). I'm not sure how VS Code picks up environment variables, but it's a lot easier to set config settings than environment variables.

jcogs33 replied:
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/561cbc4b37f6a41fbf94b58e1899653bc25654e8.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+/**
+ * Downloads the source code of a public commit from a GitHub repository.
+ */
+async function downloadPublicCommitSource(
```
Note: this function was mostly written by Copilot with DCA's `downloadPublicCommitSource` function as context.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+/**
+ * Executes the autofix command.
+ */
+function execAutofix(
```
Note: this function is mostly the same as DCA's `execFollow` function which DCA uses to execute autofix.

koesie10 commented on extensions/ql-vscode/src/view/variant-analysis/VariantAnalysisActions.tsx:
```diff
@@ -9,13 +10,17 @@ export type VariantAnalysisActionsProps = {
   hasSelectedRepositories?: boolean;
   hasFilteredRepositories?: boolean;
+
+  userSettings: UserSettings;
```
Could this just be something like `showViewAutofixesButton` to make this slightly easier to test (since you won't need to construct the full user settings)?

jcogs33 replied:
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/04a858affa7fc4b6fe49362b8b7bfa4895fa3794.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+  await withProgress(
+    async (progress: ProgressCallback) => {
+      // Get the path to the local autofix installation.
+      progress(progressUpdate(1, 4, `Checking for local autofix installation`));
```
Since the `findLocalAutofix` function should be very fast, does this actually show up in the progress window?

jcogs33 replied:
Good point. I don't think I've seen it show up. Same for the `overrideQueryHelp` progress update. I've removed them in https://github.com/github/vscode-codeql/pull/4065/commits/7d3e403e0b2b158fa4e8732123b58600b360f295.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+  if (!localAutofixPath) {
+    throw new Error("Path to local autofix installation not found.");
+  }
+  if (!pathExists(localAutofixPath)) {
```
The `pathExists` function returns a promise, so this should be:


```suggestion
  if (!(await pathExists(localAutofixPath))) {
```

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+  }
+  const queryFilePathNoExt = join(
+    dirname(queryFilePath),
+    parse(queryFilePath).name,
```
It seems like `parse` also returns a `dir`, so to avoid calling multiple `path` functions, could we store the result of `parse` in a variable and then do `join(parsedQueryFilePath.dir, parsedQueryFilePath.name)`?

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+  // Get results directory path.
+  const repoResultsStoragePath = join(repoStoragePath, "results");
+  // Find sarif file.
+  const sarifFiles = await glob(`${repoResultsStoragePath}/**/*.sarif`);
```
I believe this should always live in `results/results.sarif`. However, can we add a method to `VariantAnalysisManager`/`VariantAnalysisResultsManager` that returns this path since those should be responsible for the result storage?

Perhaps it could even make sense to first call `variantAnalysisResultsManager.isVariantAnalysisRepoDownloaded` to ensure the SARIF file is downloaded so you can give a better error message.

jcogs33 replied:
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/1d203a4bd20c7878355fccc7f54307e4f16b32da.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+    "--source-root",
+    srcRootPath,
+    "--model",
+    "capi-dev-4o", // may fail with older versions of cocofix
```
Should this be configurable (something like `codeQL.autofix.model`)?

jcogs33 replied:
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/561cbc4b37f6a41fbf94b58e1899653bc25654e8.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+    autofixArgs,
+    {
+      cwd: workDir,
+      env: {
+        CAPI_DEV_KEY: process.env.CAPI_DEV_KEY,
+        PATH: process.env.PATH,
+      },
```
Would it make sense to allow specifying the environment variables in the config (in something like `codeQL.autofix.env`)?

jcogs33 replied:
I'm not sure I'm understanding how to specify these in the config. i.e. Are you suggesting specifying the API key for `CAPI_DEV_KEY` directly in the settings.json file (is that safe)? Or specifying the path to a local file where the API key is stored? Or something else?

koesie10 replied:
I agree that we probably don't want to add the secret in the config file, but allowing you to set custom environment variables might still be useful, for example an `ENVIRONMENT` environment variable. We could also just pass through all of `process.env` to set custom environment variables though.

I'm not sure exactly how VS Code picks up environment variables, but it seems like it's possible to specify secrets when using Codespaces: [Using secrets](https://docs.github.com/en/enterprise-cloud@latest/codespaces/managing-your-codespaces/managing-your-account-specific-secrets-for-github-codespaces).

jcogs33 replied:
I'll leave as-is for now then, and maybe re-visit this idea in my follow-up PR for Go.

koesie10 reviewed (commented):
Thanks! I've not tested this locally, but the demo you shared looks good!

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+    "--diff-style",
+    "diff", // could do "text" instead if want line of "=" between fixes
+  ];
+  const varAutofixArgs = createVarAutofixArgs(
```
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/22d642945fa16a55c34e6f456493ecc067a68b19.

jcogs33 commented on extensions/ql-vscode/src/compare/compare-view.ts:
```diff
@@ -119,6 +119,8 @@ export class CompareView extends AbstractWebview<
       t: "setUserSettings",
       userSettings: {
         shouldShowProvenance: isCanary(),
```
It's not currently supported, but there's an [issue](https://github.com/github/vscode-codeql/issues/3739) to eventually support it.

jcogs33 commented:
Thanks for the reviews @koesie10 and @aeisenberg! :bow:
I've addressed some of your comments. I'll re-request your reviews once I'm finished with them all.

aeisenberg reviewed (commented):
There's one more comment from @koesie10 that should be addressed, but after that I think it's good to go.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+
+    return checkoutDir;
+  } catch (error) {
+    const errorMessage = error instanceof Error ? error.message : String(error);
```
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/d887ca505adf9a5177645c33cfc887a717e7452b.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+      await Promise.all(inputFiles.map((file) => unlink(file)));
+    }
+  } catch (error) {
+    console.error("Error merging files:", error);
```
Updated in https://github.com/github/vscode-codeql/commit/d887ca505adf9a5177645c33cfc887a717e7452b. Throwing an error instead of using the logger.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+    // Write the formatted content back to the file
+    await writeFile(inputFile, formattedContent);
+  } catch (error) {
+    console.error("Error formatting file:", error);
```
Updated in https://github.com/github/vscode-codeql/commit/d887ca505adf9a5177645c33cfc887a717e7452b. Throwing an error instead of using the logger.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
````diff
@@ -0,0 +1,801 @@
+      "```\n\n</details>\n\n ### Notes\n - notes placeholder\n\n";
+
+    // Format the content with Markdown
+    const formattedContent = `## ${header}\n\n${frontFormatting}${content}${backFormatting}`;
````
I doubt it's likely, but I added a replace for three backticks in https://github.com/github/vscode-codeql/pull/4065/commits/b7f84fca9ef88061279dd5ab6001b6006b6ecd0e.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+import { pluralize } from "../common/word";
+import { glob } from "glob";
+import { readRepoTask } from "./repo-tasks-store";
+import { unlink, mkdtemp, readFile, writeFile } from "fs/promises";
```
Switched to fs-extras in https://github.com/github/vscode-codeql/pull/4065/commits/41a3612f59a1e70d99a57bfd9129d1bf33b6ae65. (I shouldn't have let Copilot handle these imports apparently 🙁)

jcogs33 commented:
> There's one more comment from `@koesie10` that should be addressed

@aeisenberg Which comment are you referring to? There were four things I still was planning to address: three comments related to editing parts of `downloadPublicCommitSource`, and one comment about getting the autofix CLI logs into the extension logs.

aeisenberg commented:
Oh...there might be more things that I missed. I'm out tomorrow. I think it's also a holiday in the US, so I can take another look on Monday.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+
+  try {
+    // Create a temporary directory for downloading
+    const downloadDir = await mkdtemp(join(tmpdir(), "download-source-"));
```
Done in https://github.com/github/vscode-codeql/pull/4065/commits/561cd6c34b8d74056e201c19f360b8904ffe78bf.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+
+    // Extract the tarball
+    await new Promise<void>((resolve, reject) => {
+      const process = spawn("tar", ["-xzf", tarballPath], { cwd: downloadDir });
```
Switched to using a zipball and existing unzip functionality in https://github.com/github/vscode-codeql/pull/4065/commits/561cd6c34b8d74056e201c19f360b8904ffe78bf, as [suggested by `@koesie10`](https://github.com/github/vscode-codeql/pull/4065#discussion_r2179443466).

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,801 @@
+          `Invalid empty or undefined arguments: ${args.join(" ")}`,
+        );
+      }
+      const p = spawn(bin, args, { stdio: [0, 1, 2], ...options });
```
Updated in https://github.com/github/vscode-codeql/pull/4065/commits/8ed9e6a2be0b4bd80af25a1e5ed9b05d15ad4c8e. I may have oversimplified the logic 🙈, so let me know if you want any changes. I've confirmed that stderr/stdout from the autofix process are appearing in the extension logs now.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  }
+
+  // Get the query metadata.
+  const metadata = await tryGetQueryMetadata(cliServer, queryFilePath);
```
Since the metadata is required, would it make sense to just call `cliServer.resolveMetadata` directly and let it throw an error if the metadata is not available? It seems like the `tryGetQueryMetadata` was created for cases where the query metadata is optional.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  // Limit to MAX_NUM_REPOS by slicing the array.
+  if (fullNames.length > MAX_NUM_REPOS) {
+    fullNames = fullNames.slice(0, MAX_NUM_REPOS);
+    void extLogger.log(
```
Should this perhaps be a `window.showWarningMessage` call so that the user is notified more directly?

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+    await unzipToDirectoryConcurrently(
+      archivePath,
+      outputPath,
+      progressCallback
+        ? reportUnzipProgress(`Unzipping source root...`, progressCallback)
+        : undefined,
```
I think the `progressCallback` is always defined here:

```suggestion
      reportUnzipProgress(`Unzipping source root...`, progressCallback),
```

jcogs33 replied:
Ah, thanks, copy-paste issue. Updated in https://github.com/github/vscode-codeql/pull/4065/commits/0e6aa813734acc575b118799fde32794325fddfd.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  await ensureDir(outputPath);
+
+  // Define the final checkout directory
+  const checkoutDir = join(
```
I think I'm missing something, but where do we define that the archive is extracted to this directory?

jcogs33 replied:
Ah, thanks for catching this. This got affected by my refactor. The `checkoutDir` here ends up being the same directory name that `unzipToDirConcurrently` extracts to so it was still working. Fixed in https://github.com/github/vscode-codeql/pull/4065/commits/cb4bfbc517accded65ee458452cec61257a5ce0a.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+
+    return checkoutDir;
+  } catch (error) {
+    throw new Error(
```
Should we try to remove the `checkoutDir` when something in this method fails to ensure it's re-downloaded on the next try?

jcogs33 replied:
Done in https://github.com/github/vscode-codeql/pull/4065/commits/6333c9e70d7ee8c78bf510586773d4b1ae09e6e2.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  // Get autofix binary.
+  // Switch to Go binary in the future and have user pass full path
+  // in an environment variable instead of hardcoding part here.
+  const cocofixBin = join(process.cwd(), localAutofixPath, "bin", "cocofix.js");
```
If the Go binary accepts the same parameters as cocofix, would it makes sense to already add a config setting for this and use a default of `bin/cocofix.js`?

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  const localAutofixPath = getAutofixPath();
+  if (!localAutofixPath) {
+    throw new Error(
+      "Path to local autofix installation not found. Internal GitHub access required.",
+    );
```
Would it be useful to add the name of the config setting here (either hard-coded or using `AUTOFIX_PATH.qualifiedName`)?

jcogs33 replied:
Done in https://github.com/github/vscode-codeql/pull/4065/commits/bf6d8e72b871f3068c1c89465d2ad527389d90de.

koesie10 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -0,0 +1,886 @@
+  // Get autofix model from user settings.
+  const autofixModel = getAutofixModel();
+  if (!autofixModel) {
+    throw new Error("Autofix model not found.");
```
Same here, would it be useful to add the name of the config setting?

jcogs33 replied:
Done in https://github.com/github/vscode-codeql/commit/bf6d8e72b871f3068c1c89465d2ad527389d90de.

koesie10 reviewed (approved):
Thanks, this looks good to me. I have some optional comments, but I don't think any of those are necessary.

aeisenberg reviewed (commented):
Looks generally good. I see one area where I think you can simplify things.

jcogs33 commented on extensions/ql-vscode/src/variant-analysis/view-autofixes.ts:
```diff
@@ -395,91 +422,143 @@ async function downloadPublicCommitSource(
+            resolve(undefined);
+          });
+        });
+      }
```
This code mimics what is done elsewhere in the codebase, but simplifying sounds good to me. I _think_ I've done what you're suggesting in https://github.com/github/vscode-codeql/pull/4065/commits/2c04019532395b9893ddc818bd4e4327c7e74dcc (and fixed in https://github.com/github/vscode-codeql/pull/4065/commits/1f5cc9084a643d7259b669141c9319f5f3c71180).

jcogs33 commented:
@koesie10 @aeisenberg I've responded to all new review comments. Thanks for the thorough reviews! 🙇

aeisenberg commented:
I'm fine with that.
